### ISSUE TITLE
Proper way for magic link to confirm using Devise `confirmable`

### ISSUE THREAD

**Issue by wilg** [abevoelker/devise-passwordless#58]
I have both `magic_link_authenticatable` and `confirmable`. However, I want to handle the case where the user never confirms their email, but does click the magic link. This should just mark the user as confirmed, since of course they just received the email. It doesn't appear to work this way by default, and I'm not sure where to hook in. It seems like maybe its getting caught by `confirmable` first.

**Comment by wilg** [comment]
I could not figure out how to do it the way I originally described, but I overrode the controller so that attempting to send a magic link to an unconfirmed user just sends them the confirmation email instead.

```rb
class Users::Passwordless::SessionsController < Devise::Passwordless::SessionsController
  def create
    user = resource_class.find_by(email: create_params[:email])
    if user&.confirmed?
      super
    else
      user.send_confirmation_instructions
      set_flash_message!(:notice, :magic_link_sent)
      redirect_to(after_magic_link_sent_path_for(resource), status: devise_redirect_status)
    end
  end
end
```

**Comment by iainbeeston** [comment]
I'd try something like this instead, which avoids duplicating the code in the passwordless controllers 

```
class Users::MagicLinksController < Devise::MagicLinksController
  def show
    super do |resource|
      resource.confirm unless resource.confirmed?
    end
  end
end
```

**Comment by wilg** [comment]
@iainbeeston Unfortunately, that does not work for me. Devise catches it before it gets to the controller code and flashes "You have to confirm your email address before continuing." Digging around in the internals it seems quite difficult to change this behavior.

**Comment by abevoelker** [comment]
not at my dev machine to test this but try something like this

```ruby
class User
  def after_magic_link_authentication
    unless user.confirmed?
      self.skip_confirmation!
      self.save!
    end
  end
end
```

**Comment by abevoelker** [comment]
(Copying and pasting this to all open issues/PRs:)

Hey all, per #64 I unfortunately won't have much time for the foreseeable future to maintain devise-passwordless to fix the open bugs and work on new features. I'm not abandoning this project, but due to some life issues it's just at the bottom of my priority list for now.

Anyone who wants to step up and be a maintainer to shepherd the project forward would be welcomed! I just ask that you've opened a PR, or written an issue, or can otherwise demonstrate some familiarity/competence with the project. You can reply to #64 or message me privately (through email or socials since GitHub doesn't have DMs) if interested. Thank you :v:

**Comment by jasonshah** [comment]
This worked for me:

```
  # User authed with magic link. Confirm them if they are not already confirmed.
  def after_magic_link_authentication
    unless self.confirmed?
      skip_confirmation!
      self.confirmed_at = Time.now.utc
      save!
    end
  end
```

(based on @abevoelker 's [comment](https://github.com/devise-passwordless/devise-passwordless/issues/58#issuecomment-2084381804) above)

**Comment by IvanParraga** [comment]
@wilg you're right that a magic link click is exactly the same property `:confirmable` is trying to verify — control of the mailbox. They're two independent modules that don't coordinate, which is why `:confirmable` keeps blocking even after a successful magic-link verification.

The conceptual fix is to acknowledge the equivalence explicitly by calling `confirm` on the user once the magic link has verified:

```ruby
user.confirm unless user.confirmed?
sign_in(user)
```

`.confirm` is provided by `:confirmable` itself — it sets `confirmed_at` and clears any pending confirmation token. Calling it manually tells `:confirmable` "yes, mailbox control has been proven, even via a different door."

One caveat about integration: I implemented my own magic-link flow on Rails 7.1+ `generates_token_for` rather than using `devise-passwordless`, so my snippet lives in my own custom controller where I have direct control of the verify path. With `devise-passwordless` the timing is harder because `:confirmable` blocks inside the Warden strategy (via `active_for_authentication?`), before `Devise::MagicLinksController#show` ever finishes — so simply subclassing the controller and calling `confirm` after `super` doesn't work; the strategy fails first.

The cleanest long-term answer is probably a callback hook on the magic-link strategy itself (might be worth a separate issue/PR). Workarounds exist (manipulating `allow_unconfirmed_access_for` + a `Warden::Manager.after_authentication` callback scoped to the magic-link strategy), but I haven't validated those in this gem specifically.

I've written a [post]( https://www.ivanparraga.com/notes/magic-link-auth-in-rails-devise-six-production-fires-i-put-out/#fire-2-the-confirmation-step-was-redundant) walking through this exact issue (an some others related to magic links).
